### PR TITLE
datum:support json

### PR DESCRIPTION
**Pull request by AndreMouche**
Hi, 
     This PR init `json` in `datum`.  The related PR in `tidb` are  (https://github.com/pingcap/tidb/pull/3307),(https://github.com/pingcap/tidb/pull/3248).  And the related source file is [datum.go](https://github.com/pingcap/tidb/blob/master/util/types/datum.go)

@hicqu @andelf  @hhkbp2  @lishihai9017  PTAL

### CODE REVIEW

File: src/util/codec/datum.rs [hunk 1]
```diff
@@ -153,7 +166,7 @@ impl Datum {
 
     fn cmp_f64(&self, ctx: &EvalContext, f: f64) -> Result<Ordering> {
         match *self {
-            Datum::Null | Datum::Min => Ok(Ordering::Less),
+            Datum::Null | Datum::Min | Datum::Json(_) => Ok(Ordering::Less),
```
Comment (hhkbp2) [inline src/util/codec/datum.rs]: It should convert `f` to `Datum::Json()` and do the comparison.

**Comment by hhkbp2** [comment]
Rest LGTM

File: src/util/codec/datum.rs (hunk 1 above)
Comment (AndreMouche) [inline src/util/codec/datum.rs]: It seems `self` wouldn't be `Datum::Json` here.  And for this case in `tidb` it returns less(https://github.com/pingcap/tidb/blob/master/util/types/datum.go#L496), I'm not sure whether it's ok to convert `f` to `Datum::Json`  @hicqu @hhkbp2

Comment (hhkbp2) [inline src/util/codec/datum.rs]: It should not be `Datum::Json` here. But if something is wrong(a bug maybe), the code executes this branch. It should handle properly and silently by converting `f`, or panic and then crash wildly. The former is preferred.

**Comment by AndreMouche** [comment]
@hicqu @hhkbp2 @andelf @lishihai9017 PTAL

**Comment by hhkbp2** [comment]
LGTM

File: src/util/codec/datum.rs [hunk 2]
```diff
@@ -757,6 +795,10 @@ pub trait DatumEncoder: BytesEncoder + DecimalEncoder {
                     let (prec, frac) = d.prec_and_frac();
                     try!(self.encode_decimal(d, prec, frac));
                 }
+                Datum::Json(ref j) => {
+                    try!(self.write_u8(JSON_FLAG));
+                    try!(self.encode_json(j));
```
Comment (hicqu) [inline src/util/codec/datum.rs]:
```
let bytes = self.encode_json(j);
self.encode_compact_bytes(bytes);
```
Is it better?

Comment (hicqu) [inline src/util/codec/datum.rs]: tidb codec format has changed. see: https://github.com/pingcap/tidb/pull/3524

**Comment by hicqu** [comment]
LGTM.

**Pull request merged**